src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -333,47 +321,62 @@ internal void OnSolutionBuildCompleted()
             // pause live analyzer
             using var operation = _notificationService.Start("BuildDone");
             if (_diagnosticService is DiagnosticAnalyzerService)
-                SyncBuildErrorsAndReportOnBuildCompleted(inProgressState);
+                await SyncBuildErrorsAndReportOnBuildCompletedAsync(inProgressState, cancellationToken).ConfigureAwait(false);
 
             // Mark build as complete.
             OnBuildProgressChanged(inProgressState);
-            return Task.CompletedTask;
-        }, GetApplicableCancellationToken(inProgressState));
+        });
     }
 
     /// <summary>
     /// Core method that de-dupes live and build diagnostics at the completion of build.
     /// It raises diagnostic update events for both the Build-only diagnostics and Build + Intellisense diagnostics
     /// in the error list.
     /// </summary>
-    private void SyncBuildErrorsAndReportOnBuildCompleted(InProgressState inProgressState)
+    private async Task SyncBuildErrorsAndReportOnBuildCompletedAsync(InProgressState inProgressState, CancellationToken cancellationToken)
     {
-        var solution = inProgressState.Solution;
-        var cancellationToken = inProgressState.CancellationToken;
-        var allLiveErrors = inProgressState.GetLiveErrors();
+        // Allow the queue to be canceled, or this particular item to be canceled. Because we're creating a specialized
+        // token here, we need to wrap with our own try/catch to make sure that token doesn't bubble out.
+        using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(inProgressState.CancellationToken, cancellationToken);
```

Comment:
new logic to simulate the old logic.  the queue itself can be fully canceled, as can individual items.  So we get that here by linking the queue's token and the item's token, and using that for the remainder of the operation.